java/src/main/native/src/TableJni.cpp

```diff
@@ -563,6 +563,42 @@ bool valid_window_parameters(native_jintArray const &values,
          values.size() == preceding.size() && values.size() == following.size();
 }
 
+static void build_one_column_metadata(JNIEnv *env, jobject meta_obj,
+                                      jfieldID name_id,
+                                      jmethodID children_mid,
+                                      cudf::column_metadata& out) {
+  // get column name
+  cudf::jni::native_jstring col_name(env, (jstring)env->GetObjectField(meta_obj, name_id));
+  out.name = std::string(col_name.get() == NULL ? "" : col_name.get());
+  // children
+  jobjectArray j_children_meta = (jobjectArray)env->CallObjectMethod(meta_obj, children_mid);
+  cudf::jni::native_jobjectArray<jobject> children_meta(env, j_children_meta);
+  for (int i = 0; i < children_meta.size(); i ++) {
+    cudf::column_metadata cudf_col_meta;
+    build_one_column_metadata(env, children_meta.get(i), name_id, children_mid, cudf_col_meta);
+    out.children_meta.push_back(std::move(cudf_col_meta));
+  }
+}
+
+static void build_column_metadata_from_handle(JNIEnv *env, jobjectArray meta_handle,
+                                              std::vector<cudf::column_metadata>& out_meta) {
+  native_jobjectArray<jobject> col_meta(env, meta_handle);
+  out_meta.reserve(col_meta.size());
+  // Init the column meatadata class
+  jclass col_meta_cls = env->FindClass("ai/rapids/cudf/ColumnMetadata");
+  JNI_NULL_CHECK(env, col_meta_cls, "Can not find class: ai/rapids/cudf/ColumnMetadata", );
+  jfieldID name_id = env->GetFieldID(col_meta_cls, "name", "Ljava/lang/String;");
+  jmethodID children_mid =
+    env->GetMethodID(col_meta_cls, "getChildren", "()[Lai/rapids/cudf/ColumnMetadata;");
```

Comment:
No error checking on these JVM callbacks, if these callbacks remain (I would prefer they do not).

Reply:
Nice catch. The callbacks were dropped.
Instead of callbacks into JVM, updated to use regular JNI calls to build the native column metadata.